src/AppContainer.dev.js

```diff
@@ -43,6 +43,12 @@ class AppContainer extends Component {
         error: null
       });
     }
+
+    if (nextProps.children.type !== this.props.children.type) {
```

Comment:
Let's extract this to `hasChildTypeChanged` and use it everywhere we compare children. It would include backward compat logic which we would later remove.